File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
private:
    unordered_map<int, string> num_letters = {
        {2, "abc"}, 
        {3, "def"}, 
        {4, "ghi"}, 
        {5, "jkl"}, 
        {6, "mno"},
        {7, "pqrs"},
        {8, "tuv"},
        {9, "wxyz"}
    };

    void backtrack(int step, string &digits, string current, vector<string> &result) {
        if (step == digits.size() && current.size() > 0) {
            result.push_back(current);
            return;
        }

        string letters = num_letters[digits[step] - '0'];

        for(int i = 0; i < letters.size(); ++i) {
            current.push_back(letters[i]);

            backtrack(step + 1, digits, current, result);

            current.pop_back();
        }
    }

public:
    vector<string> letterCombinations(string digits) {
        vector<string> result;
        string current;

        backtrack(0, digits, current, result);

        return result;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
           # input arguaments: cur: current combination start: index of digit in digits
        # result: gloabl var to store return result
        # base case if start == len(digits) append cur to result
        # recusive start for start loop throught the chars that digits[start] represents
        # add char to cur and do backtack and then pop
        # 

        result = []
        digit_to_chars = {
            2: ["a", "b", "c"], 
            3: ["d", "e", "f"], 
            4: ["g", "h", "i"], 
            5: ["j", "k", "l"],
            6: ["m", "n", "o"],
            7: ["p", "q", "r", "s"],
            8: ["t", "u", "v"],
            9: ["w", "x", "y", "z"],
        }

        def backtrack(cur, start):
            if start == len(digits):
                result.append("".join(cur))
                return
            
            chars = digit_to_chars[int(digits[start])]

            for char in chars:
                cur.append(char)
                backtrack(cur, start + 1)
                cur.pop()
        if not digits:
            return []
            
        backtrack([], 0)
        return result

        